diff --git a/router/diag.c b/router/diag.c
index 56eb61d..33fbcbb 100644
--- a/router/diag.c
+++ b/router/diag.c
@@ -200,7 +200,8 @@ int main(int argc, char **argv)
 			errx(1, "failed to open uart\n");
 	}
 
-	diag_usb_open("/dev/ffs-diag");
+	if (getenv("DIAG_ROUTER_ENABLE_USB"))
+		diag_usb_open("/dev/ffs-diag");
 
 	ret = diag_unix_open();
 	if (ret < 0)
diff --git a/router/diag_cntl.c b/router/diag_cntl.c
index c9170c2..d749fa1 100644
--- a/router/diag_cntl.c
+++ b/router/diag_cntl.c
@@ -332,6 +332,8 @@ void diag_cntl_send_log_mask(struct peripheral *peripheral, uint32_t equip_id)
 		return;
 	if (peripheral->cntl_fd == -1) {
 		warn("Peripheral %s has no control channel. Skipping!\n", peripheral->name);
+		return;
+	} else if (peripheral->sockets && !peripheral->cntl_open) {
 		return;
 	}
 
@@ -374,6 +376,8 @@ void diag_cntl_send_msg_mask(struct peripheral *peripheral, struct diag_ssid_ran
 		return;
 	if (peripheral->cntl_fd == -1) {
 		warn("Peripheral %s has no control channel. Skipping!\n", peripheral->name);
+		return;
+	} else if (peripheral->sockets && !peripheral->cntl_open) {
 		return;
 	}
 
@@ -438,6 +442,8 @@ void diag_cntl_send_event_mask(struct peripheral *peripheral)
 		return;
 	if (peripheral->cntl_fd == -1) {
 		warn("Peripheral %s has no control channel. Skipping!\n", peripheral->name);
+		return;
+	} else if (peripheral->sockets && !peripheral->cntl_open) {
 		return;
 	}
 
